client/DESIGN.md

```diff
@@ -0,0 +1,292 @@
+
+Additionally, note that max-age and no-cache are existing HTTP Caching headers specified in [RFC7234](http://tools.ietf.org/html/rfc7234#section-5.2.1), while the read\_timestamp is sent through the Read-TxClock header, a custom HTTP header.
+
+Below are the expected forms of requests/responses between the cache and TreodeDB server:
+
+### Requests
+
+We give examples of HTTP requests below.  For additional explanation, please see the source of these examples [OMVCC and Caching](https://forum.treode.com/t/omvcc-and-http-caching/62).
+
+#### Read Requests  
+
+The cache message to the DB includes the table id, the desired key, and headers restricting the timestamps of the value.  (We describe in the Use of HTTP Headers section how each header restricts the value we read.)
+
+HTTP read requests should follow this format:
+
+    GET /table_id_string/key_string HTTP/1.1
+    Cache-Control: max-age=[max_age], no-cache
+    Read-TxClock: [read_timestamp_in_milliseconds]
+
+* **Quetion (4)**: Did I get the format for these messages (read request, write request, read response, write response) exactly correct?
+
+Here is an example of an HTTP read request:
+
+    GET /movie/star-wars HTTP/1.1
+    Cache-Control: max-age=300, no-cache
+    Read-TxClock: 1421236153024853
+
+In this example, the user is requesting the value of key "star-wars" in the table "movie".  The value should be written before (or on) time 1421236153024853 milliseconds but not more than 300 seconds before time 1421236153024853.  "no-cache" indicates that the cache MUST make a request to the DB server to retrieve this value, even if it has a value satisfying the requirements.
+
+#### Write Requests  
+
+The cache message to the DB includes the table id and a list of operations to perform on the table.  The operations should be sent to the DB server as a Treode batch write operation.
+
+The HTTP request for write should be similar to the following:
+
+    PUT /table_id_string/key_string HTTP/1.1
+    Condition-TxClock: [condition_timestamp_in_milliseconds]
+    {
+        [JSON_value]
+    }
+
+Here is an example of the HTTP write request:
+
+    PUT /movie/star-wars HTTP/1.1
+    Condition-TxClock: 1368151663681367
+    {   "title": "Star Wars",
+        "cast": [
+            { "actor": "Mark Hamill", "role": "Luke Skywalker" },
+            { "actor": "Carrie Fisher", "role": "Princess Leia Organa" },
+            { "actor": "Harrison Ford", "role": "Han Solo" } ] 
+    }
+
+We write the given JSON content to key "star-wars" in table "movie", provided that the current value associated with the "star-wars" key has remained unchanged since the time 1368151663681367 milliseconds.
+
+* **Question(5)**: How do we format batch writes as an HTTP request?
```

Comment:
For a first pass, we are going to support write batches by wrapping the objects into a JSON array. It will be something like:

    [ { "op": "create", "table": "movie", "key": "star-wars", value={ /* movie object */ } },
      { "op": "delete", "table": "movie", "key": "stars-war" } } ]

Later, we will support other mechanisms on the server side. And then the client will have its choice.

We will at some point support request pipelining on the server side. In that case, we will send each item of the batch as a separate HTTP request. In addition to GET and PUT, HTTP defines POST (for create) and DELETE. We are also allowed to add our own methods, so we'll add HOLD. Then the above would be sent as:

    PUT /movie/star-wars HTTP/1.1
    Item: 1/2

    { /* movie object */ }

    DELETE /movie/stars-war HTTP/1.1
    Item 2/2

And, if the client can support, and any intermediate HTTP caches can support it, then that will be the best method for indicating a write batch. We will support the JSON array method above, because we can be sure that HTTP libraries in all languages will let us do it. However, HTTP proxies will not understand the message to mean that their cache contents are invalidated. We will support the pipelining form eventually, because it allows intermediate HTTP proxies to see that they need to invalidate those items in their caches.